validphys2/src/validphys/scripts/vp_unweight.py

```diff
@@ -0,0 +1,154 @@
+        for k in range(len(self.reweights)):
+            for j in range(Np):
+                condition_one = j/Np - pcum[k] >= 0
+                condition_two = pcum[k+1] - j/Np >= 0
+                if condition_one and condition_two:
+                    unweights[k] += 1
+        
+        self.unweights = unweights
+        self.unprobs = unweights/np.sum(unweights)
+
+    def effective_replicas(self, weights: np.ndarray, thresh: float = 1e-12) -> int:
+        """
+        Calculate the effective number of replicas.
+
+        Args:
+            weights (np.ndarray): Array of weights.
+            thresh (float, optional): Threshold value neglecting small weights. Defaults to 1e-12.
+
+        Returns:
+            int: Effective number of replicas.
+        """
+        N = len(weights)
+        weights = weights[weights > thresh]
+        Neff = int(np.exp(-1/N*np.sum(xlogy(weights,weights/N))))
+        return Neff
+
+    def optimize(self, thresh: float, earlystopping: bool = True):
+        """
+        Optimize the unweighting process based on entropy threshold.
+
+        Args:
+            thresh (float): Entropy threshold value.
+            earlystopping (bool, optional): Whether to stop optimization early if threshold is reached. Defaults to True.
+
+        Returns:
+            Tuple[np.ndarray, np.ndarray, int]: Tuple containing arrays of Nps, entropies, and optimal Np value.
+        """
+        Nps = np.logspace(1, np.log10(len(self.weights))+1, 50, dtype=np.int64)
+        entropies = np.zeros(len(Nps))
+        for i in tqdm(range(len(Nps))):
+            self.unweight(Nps[i])
+            entropies[i] = self.entropy(self.unprobs, self.reprobs)
+            if entropies[i] <= thresh and earlystopping:
+                loc = i
+                break
+
+        if i == len(Nps)-1:
+            try:
+                loc = np.where(entropies <= thresh)[0][0]
+            except:
+                print("Failed minimisation procedure! Defaulting to lowest entropy.")
+                loc = -1
+                
+        Nopt = Nps[loc]
+
+        return Nps, entropies, Nopt
+    
+def main(chi2, N, store = True):
+    u = Unweight(None, (N, np.sqrt(chi2)))
+    u.reweight()
+    Neff = u.effective_replicas(u.reweights)
+    u.unweight(Neff)
+
+    weights = pd.DataFrame()
+    weights["unweight"] = u.unweights
+    weights["reweight"] = u.reweights
+    weights["nrep"] = np.arange(1, len(weights)+1)
+    weights = weights.set_index("nrep", drop = True)
+
+    if store:
+        weights.to_csv("weights.csv")
+
+    return weights
+
+if __name__ == "__main__":
+    parser = argparse.ArgumentParser(description="Unweighting using chi squared values.")
+    parser.add_argument("chi2_name", help = "Add the filename of the chi2 dataset (.csv)")
+    parser.add_argument("N", help = "Add the amount of experimental datapoints that the chi2 is based on")
+    args = parser.parse_args()
+    chi2 = pd.read_csv(args.chi2_name).values
```

Comment:
How is the csv generated? This should be part of the script.

Ideally the script should have two steps

1) Check whether the csv exist and if It doesn't generate it

2) Do the actual computation

In the validphys language these would be two actions, where the second depends on the first, but we can make it into proper validphys actions at the end, it can remain a script for the time being.

For the first action you can use a lot of vp stuff (e.g., functions to compute the chi2, or load pdfs).

Reply:
This I could not find a better solution for. The calculation of the chi2 per replica is done in the nnpdf/external/ poljet code and if I would shift the calculation of the chi2 to validphys, I'm afraid I must migrate more functionality. However, I don't think the solution will be more elegant.